pkg/cmd/server/server.go

```diff
@@ -0,0 +1,172 @@
+	// TODO consider an AddFlags() method on our options
+	// struct. Will need to import pflag.
+	//
+	// repeated pattern seems like it should be refactored if all
+	// options were of an interface type that specified AddFlags.
+	options.GenericServerRunOptions.AddUniversalFlags(flags)
+	options.SecureServingOptions.AddFlags(flags)
+	options.EtcdOptions.AddFlags(flags)
+	options.AuthenticationOptions.AddFlags(flags)
+	options.AuthorizationOptions.AddFlags(flags)
+
+	return cmd
+}
+
+// runServer is a method on the options for composition. allows embedding in a higher level options as we do the etcd and serving options.
+func (serverOptions ServerOptions) runServer() error {
+	fmt.Println("set up the server")
+	// options
+	// runtime options
+	if err := serverOptions.GenericServerRunOptions.DefaultExternalAddress(serverOptions.SecureServingOptions, nil); err != nil {
+		return err
+	}
+	// server configuration options
+	fmt.Println("set up serving options")
+	if err := serverOptions.SecureServingOptions.MaybeDefaultWithSelfSignedCerts(serverOptions.GenericServerRunOptions.AdvertiseAddress.String()); err != nil {
+		fmt.Printf("Error creating self-signed certificates: %v", err)
+		return err
+	}
+
+	// etcd options
+	fmt.Println("set up etcd options, do you have `--etcd-servers localhost` set?")
+	if errs := serverOptions.EtcdOptions.Validate(); len(errs) > 0 {
+		return errs[0]
+	}
+
+	// config
+	fmt.Println("set up config object")
+	genericconfig := genericapiserver.NewConfig().ApplyOptions(serverOptions.GenericServerRunOptions)
+	// these are all mutators of each specific suboption in serverOptions object.
+	// this repeated pattern seems like we could refactor
+	if _, err := genericconfig.ApplySecureServingOptions(serverOptions.SecureServingOptions); err != nil {
+		fmt.Println(err)
+		return err
+	}
+
+	// need to figure out what's throwing the `missing clientCA file` err
+	/*
+		if _, err := genericconfig.ApplyDelegatingAuthenticationOptions(serverOptions.AuthenticationOptions); err != nil {
+			fmt.Println(err)
+			return err
+		}
+	*/
+	// having this enabled causes the server to crash for any call
+	/*
+		if _, err := genericconfig.ApplyDelegatingAuthorizationOptions(serverOptions.AuthorizationOptions); err != nil {
+			fmt.Println(err)
+			return err
+		}
+	*/
+
+	// configure our own apiserver using the preconfigured genericApiServer
+	config := apiserver.Config{
+		GenericConfig: genericconfig,
+	}
+
+	// finish config
+	completedconfig := config.Complete()
+
+	// make the server
+	fmt.Println("make the server")
+	server, err := completedconfig.New()
+	if err != nil {
+		return err
+	}
+
+	// I don't like this. We're reaching in too far to call things.
+	preparedserver := server.GenericAPIServer.PrepareRun() // post api installation setup? We should have set up the api already?
+
+	stop := make(chan struct{})
+	fmt.Println("run the server")
+	preparedserver.Run(stop)
+	return nil
+}
+
+/*
+type restOptionsFactory struct {
```

Comment:
normally I'd ask you to remove it but since there are so many other temporary comments/things - we'll catch it later :-)